### PR TITLE
Add LimitStore (#2175)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #2175.

# Rationale for this change
 

# What changes are included in this PR?

Adds LimitStore

# Are there any user-facing changes?

No, a future PR will remove this functionality from the AWS implementation

### CODE REVIEW

File: object_store/src/aws.rs [hunk 1]
```diff
@@ -684,6 +684,7 @@ impl AmazonS3Builder {
 
     /// Sets the maximum number of concurrent outstanding
     /// connectons. Default is `16`.
+    #[deprecated(note = "use LimitStore instead")]
```
Comment (tustvold) [inline object_store/src/aws.rs:687]: This is a nice side-effect of #2204, we can deprecate individual things now

File: object_store/src/limit.rs [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+use std::ops::Range;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::io::AsyncWrite;
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+
+/// Store wrapper that wraps an inner store and limits the maximum number of concurrent
+/// object store operations. Where each call to an [`ObjectStore`] member function is
+/// considered a single operation, even if it may result in more than one network call
```
Comment (tustvold) [inline object_store/src/limit.rs:37]: It is worth highlighting the implementation of this functionality for S3 has the same property, despite calling it a connection limit

File: object_store/src/limit.rs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        Self {
+            inner,
+            semaphore: Arc::new(Semaphore::new(max_requests)),
+        }
+    }
+}
+
+impl<T: ObjectStore> std::fmt::Display for LimitStore<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "LimitStore({})", self.inner)
```
Comment (alamb) [inline object_store/src/limit.rs:56]: including `max_requests` would also be helpful here

File: object_store/src/limit.rs [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    }
+}
+
+#[async_trait]
+impl<T: ObjectStore> ObjectStore for LimitStore<T> {
+    async fn put(&self, location: &Path, bytes: Bytes) -> Result<()> {
+        let _permit = self.semaphore.acquire().await.unwrap();
```
Comment (alamb) [inline object_store/src/limit.rs:63]: I double checked if we needed to check the error here -- https://docs.rs/tokio/1.20.1/tokio/sync/struct.Semaphore.html#method.acquire says that errors are only returned if the Semaphore is closed, which this code does not do 👍

File: object_store/src/limit.rs [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+        let mut streams = Vec::with_capacity(max_requests);
+        for _ in 0..max_requests {
+            let stream = integration.list(None).await.unwrap();
+            streams.push(stream);
+        }
+
+        let t = Duration::from_millis(1);
+
+        // Expect to not be able to make another request
```
Comment (alamb) [inline object_store/src/limit.rs:241]: 👍 I am always a little worried when I see a `sleep` style thing in a test as I worry about intermittent failures due to timing differences, but this use seems sound 👍 

Maybe we can bump the timeout up to 20ms or something to be safe on slow machines

**Review by alamb — APPROVED** [review]
LGTM -- thanks @tustvold

**Comment by alamb** [comment]
Screen Shot 2022-07-31 at 7 09 19 AM
As an aside, if I may toot my own horn, the changes from #2149 seem to be working well and this PR only runs a small subset of the total CI checks :bowtie:

File: object_store/src/limit.rs (hunk 5 above)
Comment (tustvold) [reply object_store/src/limit.rs:241]: FWIW as this is a timeout all a slow machine could potentially mean is that the test erroneous passes, when it should have failed. I'll bump the timeout to be on the safe side

**Pull request merged**